### PR TITLE
Bounce averaged kinetic equation

### PR DESCRIPTION
**Pull request by Embreus**
Time for another big merge, I think!

The main feature of this branch is the implementation of the bounce-averaged kinetic equation, using the new PXiInternalTrapping boundary condition as well as the AnalyticB radial grid. 

It also features a large number of smaller related or unrelated tweaks and features.

Before merging, it remains to fix the 2kinetic test which now crashes. I do not yet understand why, but changing the if-continue statement on lines 151-152 in PXiExternalLoss to the one on 157-158 makes it succeed (but destabilizes the solution so that the physics test fails instead).

### CODE REVIEW

**Review by Embreus — COMMENTED** [review]
Had a pass through all changes, and left some comments.

File: examples/transport-rr/generate_consistent.py [hunk 1]
```diff
@@ -39,7 +39,7 @@
 pMax = 1    # maximum momentum in units of m_e*c
 Np   = 150  # number of momentum grid points
 Nxi  = 6    # number of pitch grid points
-tMax = 2e-3 # simulation time in seconds
+tMax = 1e-3 # simulation time in seconds
```
Comment (Embreus) [inline examples/transport-rr/generate_consistent.py:42]: this example fails with tMax=2e-3 because the nonlinear solver does not converge. I suspect that the reasion is the missing off-diagonal jacobians in the external boundary condition, but I'm not entirely sure. You end up with weak steady oscillations in the currents and E-field that do not decrease in amplitude with iterations.

File: fvm/Grid/AnalyticBRadialGridGenerator.cpp [hunk 2]
```diff
@@ -80,176 +125,269 @@ bool AnalyticBRadialGridGenerator::Rebuild(const real_t, RadialGrid *rGrid) {
...
+    real_t cdt = 1+delta_f[ir]*cosTheta;
+    real_t JOverRr = normalizedJacobian_f(ir,theta);
+    return (kappa_f[ir]*kappa_f[ir] * cosTheta * cosTheta + cdt * cdt 
+                * sdt * sdt)  / (JOverRr*JOverRr); 
+}
+
+
+void AnalyticBRadialGridGenerator::EvaluateGeometricQuantities(const len_t ir, const real_t theta, real_t &B, real_t &Jacobian, real_t &ROverR0, real_t &NablaR2){
+    real_t ct = cos(theta);
+    real_t st = sin(theta);
```
[the remark this answers is not in the document]
Comment (Embreus) [reply fvm/Grid/AnalyticBRadialGridGenerator.cpp:215]: or, rather: I wonder if there are any good routines out there that evaluate both simultaneously. I'm looking through the discussions at 
https://stackoverflow.com/questions/2683588/what-is-the-fastest-way-to-compute-sin-and-cos-together
and it sounds like there are multiple options, but they also mention that the compiler might fix this itself with the right flags. Any insights to share?

[the remark this answers is not in the document]
Comment (Embreus) [reply fvm/Grid/AnalyticBRadialGridGenerator.cpp:215]: for now I imported <math.h> and used the sincos function, which gave a huge performance boost.

File: fvm/Grid/BounceAverager.cpp [hunk 3]
```diff
@@ -240,92 +259,163 @@ real_t BounceAverager::CalculateBounceAverage(len_t ir, len_t i, len_t j, fluxGr
...
+    // XXX: assumes P-Xi grid
+    if(Bmin!=Bmax && fluxGridType==FLUXGRIDTYPE_DISTRIBUTION){
+        real_t xi_f1 = GetXi0(ir,i,j,FLUXGRIDTYPE_P2);
+        real_t xi_f2 = GetXi0(ir,i,j+1,FLUXGRIDTYPE_P2);
+        real_t xiT = grid->GetRadialGrid()->GetXi0TrappedBoundary(ir);
+        // if the cell contains, or is near, the trapped-passing boundaries, 
+        // perform more careful (and computationally expensive) cell average
+        real_t d1=fabs(xi_f1)-xiT;
+        real_t d2=fabs(xi_f2)-xiT;
+        bool containsTrappedBoundary = (xi_f1<xiT && xi_f2 >= xiT) || (xi_f1<=-xiT && xi_f2 >-xiT);
```
Comment (Embreus) [inline fvm/Grid/BounceAverager.cpp:333]: equivalently d1*d2<0 ?

File: fvm/Grid/BounceAverager.cpp [hunk 4]
```diff
@@ -240,92 +259,163 @@ real_t BounceAverager::CalculateBounceAverage(len_t ir, len_t i, len_t j, fluxGr
...
+        bool isNearTrappedBoundary = min(fabs(d1),fabs(d2)) < NEAR_TRAPPED_BOUNDARY_THRESHOLD;
+        if(containsTrappedBoundary || isNearTrappedBoundary){
+            real_t p = grid->GetMomentumGrid(ir)->GetP1(i);
+            return p*p*EvaluateCellAveragedBounceIntegralOverP2(ir,xi_f1,xi_f2,fluxGridType,F, Flist);
+        
+        // Treat singular xi0=0 case in inhomogeneous magnetic fields.
+        // (unless the cell contains the trapping region, in which 
+        //  case it would have been treated by the previous block)
+        } else if(fabs(xi0)<100*realeps){
```
Comment (Embreus) [inline fvm/Grid/BounceAverager.cpp:342]: this should rather be something like xi_f1 < 0

File: fvm/Grid/BounceAverager.cpp [hunk 5]
```diff
@@ -240,92 +259,163 @@ real_t BounceAverager::CalculateBounceAverage(len_t ir, len_t i, len_t j, fluxGr
...
+    real_t Bmax = fluxSurfaceAverager->GetBmax(ir,fluxGridType);
+
+    real_t SingularPointCorrection = 1;
+
+    // in inhomogeneous magnetic fields, for the cell centers, treat 
+    // the trapped-passing boundary and origin more carefully
+    // XXX: assumes P-Xi grid
+    if(Bmin!=Bmax && fluxGridType==FLUXGRIDTYPE_DISTRIBUTION){
+        real_t xi_f1 = GetXi0(ir,i,j,FLUXGRIDTYPE_P2);
+        real_t xi_f2 = GetXi0(ir,i,j+1,FLUXGRIDTYPE_P2);
```
Comment (Embreus) [inline fvm/Grid/BounceAverager.cpp:327]: we assume xi_f2>xi_f1. Handle opposite?

**Comment by Embreus** [comment]
I made the flux surface averager test a bit more rigorous, and discovered numerous bugs in the bounce averages. Will take some tinkering to fix, I'll look into it over the coming week.

File: fvm/Grid/AnalyticBRadialGridGenerator.cpp [hunk 6]
```diff
@@ -24,30 +24,75 @@ using namespace DREAM::FVM;
...
-    this->GsProvided     = Gs;
-    this->psisProvided   = psi_p0s;
-    this->kappasProvided = kappas;
-    // note: deltas and Deltas should vanish at rProfiles = 0 (physical constraint, fine numerically regardless)
-    this->deltasProvided = deltas;
-    this->DeltasProvided = Deltas;
-    this->nrProfiles     = nrProfiles;
-    this->rProfilesProvided = rProfiles;
+
+    // Find longest vector in 'profiles'
```
[the remark this answers is not in the document]
Comment (hoppe93) [reply fvm/Grid/AnalyticBRadialGridGenerator.cpp:35]: I think it's a left-over from an earlier implementation of the profile interpolation. Could be removed.

File: fvm/Grid/AnalyticBRadialGridGenerator.cpp [hunk 7]
```diff
@@ -24,30 +24,75 @@ using namespace DREAM::FVM;
...
+        gsl_spline *s = gsl_spline_alloc(tp, n);
+        gsl_spline_init(s, r, x, n);
+
+        return s;
+    };
+
+    // Allocate splines for shape parameters (if necessary)
+    if (pp->nG > 1) {
+        this->spline_G = construct_spline(pp->nG, pp->G_r, pp->G);
+        this->gsl_acc_G = gsl_interp_accel_alloc();
```
[the remark this answers is not in the document]
Comment (hoppe93) [reply fvm/Grid/AnalyticBRadialGridGenerator.cpp:54]: No, because we allow different shape parameters to be given on different radial grids. GSL requires all profiles using the same spline to have exactly the same size.

File: fvm/Grid/AnalyticBRadialGridGenerator.cpp [hunk 8]
```diff
@@ -24,30 +24,75 @@ using namespace DREAM::FVM;
...
+    }
+    if (pp->npsi > 1) {
```
[the remark this answers is not in the document]
Comment (hoppe93) [reply fvm/Grid/AnalyticBRadialGridGenerator.cpp:56]: Sounds like a good idea!

File: fvm/Grid/BounceAverager.avalancheDelta.cpp [hunk 9]
```diff
@@ -0,0 +1,228 @@
+/**
```
[the remark this answers is not in the document]
Comment (hoppe93) [reply fvm/Grid/BounceAverager.avalancheDelta.cpp:1]: I think it's a good idea to split the file!

File: fvm/Grid/AnalyticBRadialGridGenerator.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (hoppe93) [reply fvm/Grid/AnalyticBRadialGridGenerator.cpp]: I know that *Numerical Recipes* have a discussion about how to compute sin and cos efficiently, and I also think they make some wild statement like "if computing sin and cos is where most of the time is spent, something is seriously wrong with your code". I don't recall now what their method is, and the book is at the office so I can't check :P

File: include/FVM/Grid/RadialGrid.hpp [hunk 10]
```diff
@@ -187,10 +195,12 @@ namespace DREAM::FVM {
          * Returns q*R0 on the distribution grid where q 
          * is the safety factor and R0 the major radius.
          *  ir: radial grid index
-         *  mu0Ip: product of vacuum permeability and plasma 
+         *  mu0Ip: product of vacuum permeability and toroidal plasma 
          *         current enclosed by the flux surface ir. 
          */
```
[the remark this answers is not in the document]
Comment (hoppe93) [reply include/FVM/Grid/RadialGrid.hpp:200]: I think it's the right place to put it. The safety factor is a quantity which we should be able to evaluate regardless of the geometry (I guess, at least for geometries trying to mimic a tokamak; maybe not so much for a coronal loop...)

File: fvm/Grid/AnalyticBRadialGridGenerator.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Embreus) [reply fvm/Grid/AnalyticBRadialGridGenerator.cpp:54]: but isn't the gsl_acc object independent of the spline?

[the remark this answers is not in the document]
Comment (hoppe93) [reply fvm/Grid/AnalyticBRadialGridGenerator.cpp:54]: Not really. The ``gsl_acc`` keeps track of where the spline has been evaluated earlier and which grid points that corresponds to. If we switch the grid which the accelerator is working with all the time we will confuse it and lose the advantage of the accelerator.

[the remark this answers is not in the document]
Comment (Embreus) [reply fvm/Grid/AnalyticBRadialGridGenerator.cpp:54]: Oh, I had no clue! I may be using it inappropriately elsewhere; will review.

**Comment by Embreus** [comment]
I would like to make the merge now. I suggest we change the 2kinetic test to only utilize a cylindrical grid for now so that tests check out, open a separate issue for the KineticKinetic boundary condition with trapping effects, and pull to master. There are so many improvements on this branch now that should be of interest to all users, for example some of Ida's simulations are ~3 times faster on this branch.
